pkg/backup/backup.go

```diff
@@ -174,94 +175,186 @@ func (b *Backup) Run() {
+		return lastSnapRev, nil
 	}
 
-	if len(pods) == 0 {
-		msg := "no running etcd pods found"
-		logrus.Warning(msg)
-		return lastSnapRev, fmt.Errorf(msg)
-	}
-	member, rev := getMemberWithMaxRev(pods, b.etcdTLSConfig, b.selfHosted)
-	if member == nil {
-		logrus.Warning("no reachable member")
-		return lastSnapRev, fmt.Errorf("no reachable member")
+	etcdVersion, err := snap.Version()
+	if err != nil {
+		return lastSnapRev, fmt.Errorf("failed to get etcd version: (%v)", err)
 	}
 
-	if rev <= lastSnapRev {
-		logrus.Info("skipped creating new backup: no change since last time")
-		return lastSnapRev, nil
+	n, err := WriteSnap(etcdVersion, rev, b.be, snap)
+	if err != nil {
+		return lastSnapRev, fmt.Errorf("failed to write snapshot to backend: (%v)", err)
 	}
 
-	log.Printf("saving backup for cluster (%s)", b.clusterName)
-	if err := b.writeSnap(member, rev); err != nil {
-		err = fmt.Errorf("write snapshot failed: %v", err)
-		return lastSnapRev, err
+	bs := backupapi.BackupStatus{
+		CreationTime:     time.Now().Format(time.RFC3339),
+		Size:             util.ToMB(n),
+		Version:          etcdVersion,
+		Revision:         rev,
+		TimeTookInSecond: int(time.Since(start).Seconds() + 1),
 	}
+	b.recentBackupsStatus = append(b.recentBackupsStatus, bs)
+	if len(b.recentBackupsStatus) > maxRecentBackupStatusCount {
+		b.recentBackupsStatus = b.recentBackupsStatus[1:]
+	}
+
 	return rev, nil
 }
 
-func (b *Backup) writeSnap(m *etcdutil.Member, rev int64) error {
-	start := time.Now()
+// WriteSnap writes the snapshot to the backend.
+func WriteSnap(version string, rev int64, be backend.Backend, snap *Snap) (int64, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), constants.DefaultSnapshotTimeout)
+	defer cancel()
+	rc, err := snap.Snapshot(ctx)
+	if err != nil {
+		return 0, fmt.Errorf("failed to get snapshot (%v)", err)
+	}
+
+	n, err := be.Save(version, rev, rc)
+	rc.Close()
+	if err != nil {
+		return 0, err
+	}
+	return n, nil
+}
+
+// GetSnap gets the snapshot from the cluster member with the maximum revision.
+func GetSnap(kclient kubernetes.Interface, tls *tls.Config, namespace, clusterName string) (*Snap, int64, error) {
+	pods, err := getRunningPods(kclient, namespace, clusterName)
+	if err != nil {
+		return nil, 0, fmt.Errorf("failed to retrieve running etcd pods: (%v)", err)
+	}
+	if len(pods) == 0 {
+		return nil, 0, fmt.Errorf("no running etcd pods found")
+	}
+
+	m, rev := getMemberWithMaxRev(pods, tls)
+	if m == nil {
+		return nil, 0, fmt.Errorf("no reachable member")
+	}
+
+	return NewSnap(m.ClientURL(), tls), rev, nil
+}
 
+func NewSnap(url string, tls *tls.Config) *Snap {
 	cfg := clientv3.Config{
-		Endpoints:   []string{m.ClientURL()},
+		Endpoints:   []string{url},
 		DialTimeout: constants.DefaultDialTimeout,
-		TLS:         b.etcdTLSConfig,
+		TLS:         tls,
 	}
-	etcdcli, err := clientv3.New(cfg)
+	return &Snap{cfg: cfg}
+}
+
+type Snap struct {
```

Comment:
what does snap represent for? needs doc string

Reply:
Snap provides an abstraction of getting snapshot from a member. 
 so calling snap.snapshot() returns the snapshot. calling snap.close() releases resources which is the etcd client.

Reply:
Then it should be named as SnapshotGetter. I am not sure if this abstraction is really necessary though.

Reply:
let me rethink about this.